### PR TITLE
Allow add_explicit_type to replace a placeholder type

### PR DESCRIPTION
**Pull request by lnicola**

### CODE REVIEW

File: crates/ra_assists/src/assists/add_explicit_type.rs [hunk 1]
```diff
@@ -34,17 +35,29 @@ pub(crate) fn add_explicit_type(ctx: AssistCtx<impl HirDatabase>) -> Option<Assi
...
     if !cursor_in_range {
         return None;
     }
     // Assist not applicable if the type has already been specified
-    if stmt.syntax().children_with_tokens().any(|child| child.kind() == T![:]) {
-        return None;
+    // and it has no placeholders
+    if ty_range.is_some() {
+        if stmt.syntax().children_with_tokens().any(|child| child.kind() == T![:]) {
+            if !stmt.syntax().descendants().any(|child| child.kind() == PLACEHOLDER_TYPE) {
```
Comment (matklad) [inline crates/ra_assists/src/assists/add_explicit_type.rs]: I think we should walk only the type , and not the whole statement here (which includes the initalizing expression). 

it seems like the following should do the trick
```
match stmt.ascribed_type() {
    None => /* no type at all, try to insert one */,
    Some(ty) => /* there's already a type, let's check if it contains `_` */
}
```

File: crates/ra_assists/src/assists/add_explicit_type.rs [hunk 2]
```diff
@@ -85,6 +101,24 @@ mod tests {
...
+            "fn f() { let a<|>: i32 = 1; }",
+        );
+    }
+
+    #[test]
+    fn add_explicit_type_works_for_nested_underscore() {
+        check_assist(
+            add_explicit_type,
+            "fn f() { let a<|>: Option<_> = Some(1); }",
+            "fn f() { let a<|>: Option<i32> = Some(1); }",
```
Comment (matklad) [inline crates/ra_assists/src/assists/add_explicit_type.rs]: Yeah, this test would need an `Option` defined.

**Comment by lnicola** [comment]
No worries if you didn't get to look at this. Just checking that it didn't slip through the cracks.

**Pull request merged**